lib/network_utils.ml

```diff
@@ -1163,3 +1200,196 @@ module Ethtool = struct
+				| n when n <= 0 -> acc
+				| n -> aux (value :: acc) (n-1)
+			in
+			aux [] repeat
+		in
+		Sysfs.get_dev_nums_with_same_driver driver
+		|> gen_list (string_of_int max_vfs)
+		|> String.concat ","
+
+	(* For given driver like igb, we parse each line of igb.conf which is the modprobe
+	configuration for igb. We keep the same the lines that do not have SRIOV configurations and 
+	change lines that need to be changed with patterns like `options igb max_vfs=4`
+	*)
+	let parse_modprobe_conf_internal file_path driver option =
+		(* Initially I did not use ref here, but afterward changed to ref to make the code easy to read *)
+		let has_probe_conf = ref false in
+		let need_rebuild_initrd = ref false in
+		let parse_single_line s = 
+			let parse_driver_options s = 
+				match Xstringext.String.split ~limit:2 '=' s with
+				(* has SRIOV configuration but the max_vfs is exactly what we want to set, so no changes and return s *)
+				| [k; v] when k = "max_vfs" && v = option ->  has_probe_conf := true; s
+				(* has SRIOV configuration and we need change it to expected option *)
+				| [k; v] when k = "max_vfs"  -> 
+					has_probe_conf := true;
+					need_rebuild_initrd := true;
+					debug "change SRIOV options from [%s=%s] to [%s=%s]" k v k option;
+					Printf.sprintf "max_vfs=%s" option
+				(* we do not care the lines without SRIOV configurations *)
+				| _ -> s
+			in
+			let trimed_s = String.trim s in
+			if Re.execp (Re_perl.compile_pat ("options[ \t]+" ^ driver)) trimed_s then
+				let driver_options = Re.split (Re_perl.compile_pat "[ \t]+") trimed_s in
+				List.map parse_driver_options driver_options
+				|> String.concat " "
+			else
+				trimed_s
+		in
+		let lines = try Unixext.read_lines file_path with _ -> [] in
+		let new_conf = List.map parse_single_line lines in
+		!has_probe_conf, !need_rebuild_initrd, new_conf
+
+	(*
+	returns ( a * b * c) where
+	a indicates the probe configuration already has the max_vfs options, meaning the device doesn't support sysfs and will be configed by modprobe
+	b indicates some changes shall be made on the coniguration to enable sriov to max_vfs, so we shall rebuild the initrd.
+	and c is the configurations after these changes
+	*)
+	let parse_modprobe_conf driver max_vfs =
+		let file_path = Printf.sprintf "/etc/modprobe.d/%s.conf" driver in
+		let option = gen_options_for_maxvfs driver max_vfs in
+		parse_modprobe_conf_internal file_path driver option
+
+	let get_sriov_info lspci_output =
+		if Re.execp (Re_perl.compile_pat "Capabilities:.+SR-IOV") lspci_output then
+		begin
+			let group = Re.exec_opt (Re_perl.compile_pat "Total VFs: (\d+), Number of VFs: (\d+)") lspci_output in
+			match group with
+			| None -> {max_vfs=0; num_vfs=0}
+			| Some x -> {max_vfs = (int_of_string (Re.Group.get x 1)) - 1;
+				num_vfs = (int_of_string (Re.Group.get x 2))}
+		end
+		else
+			{max_vfs=0; num_vfs=0}
+
+	let get_lspci_output dev = 
+		call_script lspci ["-vv"; "-s"; (Sysfs.get_pcibuspath dev)]
```

Comment:
We should try to use `ocaml-pci` rather than calling `lspci` and parsing its output.

Reply:
Or, it may be easier to get the information from sysfs. For example, to check whether a NIC supports SR-IOV, we can look at the `sriov_totalvfs` node instead of the capabilities from lspci:

>What:		/sys/bus/pci/devices/.../sriov_totalvfs
Date:		November 2012
Contact:	Donald Dutile <ddutile@redhat.com>
Description:
		This file appears when a physical PCIe device supports SR-IOV.
		Userspace applications can read this file to determine the
		maximum number of Virtual Functions (VFs) a PCIe physical
		function (PF) can support. Typically, this is the value reported
		in the PF's SR-IOV extended capability structure's TotalVFs
		element.  Drivers have the ability at probe time to reduce the
		value read from this file via the pci_sriov_set_totalvfs()
		function.

Reply:
I remember that I didn't use sriov_totalvfs file is because it doesn't always show the correct totalvfs number. The standard should be sriov_totalvfs = sriov_maxvfs + 1 (totalvfs = pf number + vf number), but for igbxe devices, the value in file sriov_totalvfs is 63 which is exactly the same with max_vfs. 

I used lspci's output is because its output is always correct.

Reply:
Moreover, I didn't find ocaml-pci able to get the capabilities fields of PCI.

I will change to use sysfs to get sriov-totalvfs, the disadvantage of which is we will only enable 62 VFs on igbxe devices (maximum VFs is actually 63).

Reply:
Are you saying that lspci has a special case hardcoded for igbxe? Because, as far as I understand it, lspci/libpci by default also uses sysfs to get this kind of information.

For the capability checking, however, it doesn't matter whether it gets the number right. If the file exists, then the SR-IOV is present, if not, then it isn't.